### PR TITLE
Add IMEX SolveImplicitSector mutator

### PR DESCRIPTION
**Pull request by wthrowe**
This is the actual solver for the IMEX code.  The remaining stuff to submit is mostly actions and other similar wrappers.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Imex/Tags/Mode.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#pragma once
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "Evolution/Imex/Mode.hpp"
+#include "Evolution/Tags.hpp"
+#include "Options/Options.hpp"
+
+namespace imex {
+namespace OptionTags {
+struct ImexMode {
```
Comment (nilsdeppe) [inline src/Evolution/Imex/Tags/Mode.hpp]: Do you want to add Doxygen comments for the tags, mostly for discoverability?

File: src/Evolution/Imex/Tags/Mode.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  static constexpr Options::String help{"IMEX implementation to use"};
+  using type = ::imex::Mode;
+  using group = evolution::OptionTags::Group;
```
Comment (nilsdeppe) [inline src/Evolution/Imex/Tags/Mode.hpp]: Would it be useful to add an `Imex` subgroup into `Evolution` to group the options together? [later in a separate PR is totally fine if you think it's a good idea!]

File: src/Evolution/Imex/Tags/SolveTolerance.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+#pragma once
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "Evolution/Tags.hpp"
+#include "Options/Options.hpp"
+
+namespace imex {
+namespace OptionTags {
+struct ImexImplicitSolveTolerance {
```
Comment (nilsdeppe) [inline src/Evolution/Imex/Tags/SolveTolerance.hpp]: same doxygen question

File: src/Evolution/Imex/SolveImplicitSector.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+ private:
+  template <typename Attempt>
+  struct get_tags_from_evolution {
+    using type = typename Attempt::tags_from_evolution;
+  };
+
+  using attempt_tags = tmpl::transform<typename ImplicitSector::solve_attempts,
+                                       get_tags_from_evolution<tmpl::_1>>;
+  using evolution_data_tags =
+      tmpl::push_front<attempt_tags,
```
Comment (nilsdeppe) [inline src/Evolution/Imex/SolveImplicitSector.hpp]: Is it possible `tmpl::append` here to combine the two lists? If not, maybe a comment why?

File: src/Evolution/Imex/SolveImplicitSector.tpp [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+               initial_guess_arguments, tmpl::type_<ImplicitSector>{});
+  if (not equation.do_solve()) {
+    return;
+  }
+
+  const size_t number_of_grid_points = get(*solve_failures).size();
+  // Only allocated if used.
+  Matrix semi_implicit_jacobian{};
+
+  bool had_failure = true;
```
Comment (nilsdeppe) [inline src/Evolution/Imex/SolveImplicitSector.tpp]: I think I would find reading the code easier if this was `solve_succeeded` or something like that. `not had_failure` I think is `success` and I would find it easier to read
```cpp
if (solve_succeeded) {
  return;
}
```

File: src/Evolution/Imex/SolveImplicitSector.tpp [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+              had_failure = true;
+              continue;
+            } else {
+              throw;
+            }
+          }
+          break;
+        }
+        case Mode::SemiImplicit: {
+          auto correction_array = solver(initial_guess);
```
Comment (nilsdeppe) [inline src/Evolution/Imex/SolveImplicitSector.tpp]: I'm a bit unclear on what memory allocations are being done in the `SemiImplicit` solve. Does this `solver` call and the `jacobian()` call below do memory allocations? I think the `lapack::general_matrix_linear_solve` call also does a `std::vector` allocation inside. If we are doing those three allocations, could they be combine or shared between grid points? If that's already being done, would a comment or two in this part of the switch statement be useful?

Edit: from looking at the above code I think there aren't any allocations, other than maybe inside `general_matrix_linear_solve`? Maybe a few comments here would help make that clear?

File: src/Evolution/Imex/SolveImplicitSector.tpp [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+           "initial_guess must return one GuessResult per point or an empty "
+           "vector for GuessResult::InitialGuess everywhere.");
+  }
+
+  double implicit_weight() const { return implicit_weight_; }
+  const SectorVariables& inhomogeneous_terms() const {
+    return inhomogeneous_terms_;
+  }
+
+  bool do_solve() const { return implicit_weight_ != 0.0; }
```
Comment (nilsdeppe) [inline src/Evolution/Imex/SolveImplicitSector.tpp]: the function name makes me think it does some computation. Maybe rename to something like `solved_needed`?

File: src/Evolution/Imex/SolveImplicitSector.tpp [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+      db::compute_databox_type<tmpl::append<simple_tags, compute_tags>>;
+
+ public:
+  template <typename PassedEvolutionData>
+  ImplicitSolver(const ImplicitEquation<SectorVariables>& implicit_equation,
+                 PassedEvolutionData&& data_from_evolution)
+      : solve_box_(db::create<simple_tags, compute_tags>()),
+        implicit_equation_(&implicit_equation) {
+    static_assert(std::is_same_v<PassedEvolutionData, const EvolutionData&>,
+                  "ImplicitSolver was passed a temporary.");
```
Comment (nilsdeppe) [inline src/Evolution/Imex/SolveImplicitSector.tpp]: Could you expand this error message? I'm not sure what I should do if I were to get this.

File: src/Evolution/Imex/SolveImplicitSector.tpp [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+  }
+
+ private:
+  double implicit_weight_;
+  SectorVariables inhomogeneous_terms_{};
+  std::vector<GuessResult> initial_guess_types_{};
+};
+
+template <typename ImplicitSector, typename SolveAttempt>
+class ImplicitSolver {
```
Comment (nilsdeppe) [inline src/Evolution/Imex/SolveImplicitSector.tpp:137]: Would some overview documentation be useful here? Not necessarily for doxygen, but to get a quick overview of what the different functions do and why? Maybe also for `ImplicitEquation`?

File: tests/Unit/Helpers/Evolution/Imex/TestSector.hpp [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+template <typename Tag>
+struct ReadOnlySource : ::db::SimpleTag {
+  using type = typename Tag::type;
+};
+
+template <typename Tag>
+struct ReadOnly : Tag, ::db::ComputeTag {
+  using base = Tag;
+  using argument_tags = tmpl::list<ReadOnlySource<Tag>>;
+  static void function(gsl::not_null<typename ReadOnly::type*> dest,
```
Comment (nilsdeppe) [inline tests/Unit/Helpers/Evolution/Imex/TestSector.hpp]: `const gsl::not_null`?

File: tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp [hunk 11]
```diff
@@ -0,0 +1,750 @@
...
+    CHECK(not performing_step_with_no_implicit_term);
+
+    std::pair current_values{var2, var3};
+    CHECK((*prep_run_values)[static_cast<size_t>(Prep)] != current_values);
+    (*prep_run_values)[static_cast<size_t>(Prep)] = std::move(current_values);
+  }
+};
+// End stuff only used for DataBox handling
+
+// [initial_guess]
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp:138]: Do you want this `initial_guess` to be around the `AnalyticSolution` or the `InitialGuess` class below? Right now I think it's around the `AnalyticSolution`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
This is great! Thanks for all your hard work on the IMEX code! I have a few suggestions/comments/questions. Please feel free to rebase and squash directly since I don't think any of them are major.

File: src/Evolution/Imex/SolveImplicitSector.hpp (hunk 4 above)
Comment (wthrowe) [reply src/Evolution/Imex/SolveImplicitSector.hpp]: Renamed a thing and added some comments to try to make this clearer.  Not sure if it's enough.

File: src/Evolution/Imex/SolveImplicitSector.tpp (hunk 8 above)
Comment (wthrowe) [reply src/Evolution/Imex/SolveImplicitSector.tpp]: You should stop passing a temporary.  I've added a brief explanation of why.

File: src/Evolution/Imex/SolveImplicitSector.tpp (hunk 6 above)
Comment (wthrowe) [reply src/Evolution/Imex/SolveImplicitSector.tpp]: There weren't supposed to be any after the first point, but you're right that I missed one in the LAPACK wrapper, so I fixed that.

I've made the types of some variables and temporaries explicitly arrays, so it should be clear that there are no allocations in them.

File: tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp (hunk 11 above)
Comment (wthrowe) [reply tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp:138]: This is intentional.  They are both valid "initial_guess" structs (the example chooses between them with a `tmpl::conditional_t`), but this one does a reasonable calculation while `InitialGuess` just does some nonsensical modifications to the data.

File: src/Evolution/Imex/Tags/Mode.hpp (hunk 2 above)
Comment (wthrowe) [reply src/Evolution/Imex/Tags/Mode.hpp]: Done, and renamed some tags both to avoid duplicating the group name and for consistency.

**Review by wthrowe — COMMENTED** [review]
This is now dependent on (and includes the commits from) #5657.

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please go ahead and squash!

**Comment by wthrowe** [comment]
@nilsdeppe I think this is ready to go.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**